File: cpp/src/analytics/distributions.cpp
/**
 * AlgoForge — src/analytics/distributions.cpp
 * Descriptive statistics: skewness, excess kurtosis, quantiles.
 */

#include "core/analytics.hpp"

#include <algorithm>
#include <cmath>
#include <span>
#include <vector>

namespace algoforge::analytics {

double percentile_linear(std::span<const double> sorted_values, double pct) noexcept {
    const int n = static_cast<int>(sorted_values.size());
    if (n == 0) return 0.0;
    if (n == 1) return sorted_values[0];

    /* R-7 / numpy 'linear' method */
    double index = (pct / 100.0) * (n - 1);
    int    lo    = static_cast<int>(index);
    int    hi    = lo + 1;
    if (hi >= n) return sorted_values[n - 1];
    double frac  = index - lo;
    return sorted_values[lo] + frac * (sorted_values[hi] - sorted_values[lo]);
}

DistributionStats compute_distribution(std::span<const double> values) noexcept {
    DistributionStats ds;
    const int n = static_cast<int>(values.size());
    if (n == 0) return ds;

    ds.n = n;

    /* mean */
    double sum = 0.0;
    for (double v : values) sum += v;
    double mean = sum / n;
    ds.mean = mean;

    /* sample std dev, biased std dev, min, max */
    double ss = 0.0;
    double m3 = 0.0;
    double m4 = 0.0;
    ds.min_val = values[0];
    ds.max_val = values[0];

    for (double v : values) {
        double d = v - mean;
        ss += d * d;
        m3 += d * d * d;
        m4 += d * d * d * d;
        if (v < ds.min_val) ds.min_val = v;
        if (v > ds.max_val) ds.max_val = v;
    }

    double var_sample = (n > 1) ? ss / (n - 1) : 0.0;
    ds.std_dev = std::sqrt(var_sample);

    /* biased std dev (for skewness/kurtosis formulas) */
    double std_biased = std::sqrt(ss / n);

    if (n >= 4 && std_biased > 1e-15) {
        /* g1 = (1/n)*sum((x-mean)^3) / std_biased^3 */
        double g1 = (m3 / n) / (std_biased * std_biased * std_biased);
        /* Fisher-Pearson adjusted G1 */
        double G1 = (std::sqrt(static_cast<double>(n) * (n - 1)) / (n - 2)) * g1;
        ds.skewness = G1;

        /* g2 = (1/n)*sum((x-mean)^4) / std_biased^4 - 3 */
        double g2 = (m4 / n) / (std_biased * std_biased * std_biased * std_biased) - 3.0;
        /* Fisher adjusted G2 */
        double G2 = (static_cast<double>(n - 1) / (static_cast<double>(n - 2) * (n - 3)))
                    * ((n + 1) * g2 + 6.0);
        ds.excess_kurtosis = G2;
    }
    /* else skewness = 0, excess_kurtosis = 0 (default) */

    /* Quantiles — requires sorted copy */
    std::vector<double> sorted(values.begin(), values.end());
    std::sort(sorted.begin(), sorted.end());

    ds.p25 = percentile_linear(sorted, 25.0);
    ds.p50 = percentile_linear(sorted, 50.0);
    ds.p75 = percentile_linear(sorted, 75.0);

    return ds;
}

} /* namespace algoforge::analytics */

File: js/src/analytics/distributions.js
/**
 * distributions.js — skewness, excess kurtosis, quantiles, full stats.
 * Spec §4 (Distributions section).
 * Uses Fisher-Pearson adjusted skewness and Fisher's adjusted kurtosis.
 */

/**
 * Sample mean.
 * @param {number[]} xs
 * @returns {number}
 */
function mean(xs) {
    if (xs.length === 0) return 0;
    let s = 0;
    for (const x of xs) s += x;
    return s / xs.length;
}

/**
 * Sample (biased) standard deviation — denominator n.
 * @param {number[]} xs
 * @param {number}   mu  precomputed mean
 * @returns {number}
 */
function stdBiased(xs, mu) {
    if (xs.length < 2) return 0;
    let s = 0;
    for (const x of xs) { const d = x - mu; s += d * d; }
    return Math.sqrt(s / xs.length);
}

/**
 * Fisher-Pearson adjusted skewness (G1).
 * Returns 0 for n < 3.
 * @param {number[]} xs
 * @returns {number}
 */
export function skewness(xs) {
    const n = xs.length;
    if (n < 3) return 0;
    const mu = mean(xs);
    const sb = stdBiased(xs, mu);
    if (sb < 1e-14) return 0;
    let s = 0;
    for (const x of xs) { const d = (x - mu) / sb; s += d * d * d; }
    const g1 = s / n;
    return Math.sqrt(n * (n - 1)) / (n - 2) * g1;
}

/**
 * Fisher's excess kurtosis (adjusted for sample).
 * Returns 0 for n < 4.
 * @param {number[]} xs
 * @returns {number}
 */
export function excessKurtosis(xs) {
    const n = xs.length;
    if (n < 4) return 0;
    const mu = mean(xs);
    const sb = stdBiased(xs, mu);
    if (sb < 1e-14) return 0;
    let s = 0;
    for (const x of xs) { const d = (x - mu) / sb; s += d * d * d * d; }
    const g2 = s / n - 3;
    return ((n - 1) / ((n - 2) * (n - 3))) * ((n + 1) * g2 + 6);
}

/**
 * Linear interpolation quantile (R-7 / numpy method='linear').
 * @param {number[]} xs   data array (need not be sorted)
 * @param {number}   p    quantile in [0, 1]
 * @returns {number}
 */
export function quantile(xs, p) {
    const n = xs.length;
    if (n === 0) return NaN;
    if (n === 1) return xs[0];
    const sorted = xs.slice().sort((a, b) => a - b);
    const h = (n - 1) * p;
    const lo = Math.floor(h);
    const hi = Math.ceil(h);
    if (lo === hi) return sorted[lo];
    return sorted[lo] + (h - lo) * (sorted[hi] - sorted[lo]);
}

/**
 * Full distribution statistics for an array.
 * @param {number[]} xs
 * @returns {{
 *   n: number, mean: number, std: number, skewness: number,
 *   excessKurtosis: number,
 *   min: number, max: number,
 *   p5: number, p25: number, p50: number, p75: number, p95: number,
 * }}
 */
export function distributionStats(xs) {
    const n = xs.length;
    if (n === 0) {
        return {
            n: 0, mean: 0, std: 0, skewness: 0, excessKurtosis: 0,
            min: NaN, max: NaN,
            p5: NaN, p25: NaN, p50: NaN, p75: NaN, p95: NaN,
        };
    }
    const mu = mean(xs);
    // sample std (n-1)
    let sq = 0;
    for (const x of xs) { const d = x - mu; sq += d * d; }
    const sampleStd = n > 1 ? Math.sqrt(sq / (n - 1)) : 0;

    let mn = xs[0], mx = xs[0];
    for (const x of xs) { if (x < mn) mn = x; if (x > mx) mx = x; }

    return {
        n,
        mean:           mu,
        std:            sampleStd,
        skewness:       skewness(xs),
        excessKurtosis: excessKurtosis(xs),
        min:            mn,
        max:            mx,
        p5:             quantile(xs, 0.05),
        p25:            quantile(xs, 0.25),
        p50:            quantile(xs, 0.50),
        p75:            quantile(xs, 0.75),
        p95:            quantile(xs, 0.95),
    };
}
